homeassistant/components/supla/__init__.py

```diff
@@ -0,0 +1,158 @@
+"""Support for the Supla devices."""
+
+import logging
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_ACCESS_TOKEN
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['pysupla==0.0.2']
+
+
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = 'supla'
+
+CONF_SERVER = 'server'
+CONF_SERVERS = 'servers'
+
+SUPLA_FUNCTION_HA_CMP_MAP = {
+    'CONTROLLINGTHEROLLERSHUTTER': 'cover'
+}
+SUPLA_CHANNELS = 'supla_channels'
+SUPLA_SERVERS = 'supla_servers'
+
+
+SERVER_CONFIG = vol.Schema({
+    vol.Required(CONF_SERVER): cv.string,
+    vol.Required(CONF_ACCESS_TOKEN): cv.string
+}, extra=vol.ALLOW_EXTRA)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_SERVERS):
+            vol.All(cv.ensure_list, [SERVER_CONFIG])
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def discover_devices(hass, hass_config):
+    """
+    Run periodically to discover new devices.
+
+    Currently it's only run at startup.
+    """
+    component_configs = {}
+
+    for server_name, server in hass.data[SUPLA_SERVERS].items():
+
+        for channel in server.get_channels(include=['iodevice']):
+            channel_function = channel['function']['name']
+            component_name = SUPLA_FUNCTION_HA_CMP_MAP.get(channel_function)
+
+            if component_name is None:
+                _LOGGER.warning(
+                    'Unsupported function: %s, channel id: %s',
+                    channel_function, channel['id']
+                )
+                continue
+
+            channel['server_name'] = server_name
+            component_configs.setdefault(component_name, []).append(channel)
+
+    # Load discovered devices
+    for component_name, channel in component_configs.items():
+        load_platform(
+            hass,
+            component_name,
+            'supla',
+            channel,
+            hass_config
+        )
+
+
+def setup(hass, base_config):
+    """Set up the Supla component."""
+    from pysupla import SuplaAPI
+
+    server_confs = base_config[DOMAIN][CONF_SERVERS]
+
+    hass.data[SUPLA_SERVERS] = {}
+    hass.data[SUPLA_CHANNELS] = {}
+
+    for server_conf in server_confs:
+        server = SuplaAPI(
```

Comment:
We should validate that the access token works before continuing with set up.

Reply:
Validation added.